woocommerce/woocommerce
# Track frequency of unhandled JS errors with MC Stats

Track frequency of unhandled JS errors with MC Stats

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes https://github.com/woocommerce/woocommerce/issues/49289


This PR adds a new function, `bumpStat,` to the `@woocommerce/tracks` package and uses it to track the frequency of unhandled JS errors.

The implementation is based on Calypso's [`bumpStat()`](https://github.com/Automattic/wp-calypso/blob/31d3844d216c261ed9bff5ea0abadc0aa0d78b4b/client/lib/analytics/mc.js#L34-L50). 

Reference: PCYsg-J0-p2



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

To test this PR, we need to use the local dev environment and follow the below steps to simulate the error and check if it is logged in the server.

1. Checkout this PR branch.
2. Run `pnpm install`
3.  Install Code Snippets plugin and add a new snippet with the below code to enable the remote logging.

		```php
		update_option( 'woocommerce_feature_remote_logging_enabled', 'yes' );
		update_option( 'woocommerce_allow_tracking', 'yes' );

		add_filter(
			'option_woocommerce_remote_variant_assignment',
			function () {
				return 5;
			}
		);

4. Go to `WooCommerce > Home` page.
5. Open the browser dev tools and go to the console tab.
6. Run `window.localStorage.setItem( 'debug', '*')` to enable the debug logging.
7. Now, run `git cherry-pick 146a650` (146a650) to simulate the error.
8. Run `pnpm --filter='@woocommerce/plugin-woocommerce' build` to build production files.
9. Confirm `wc-admin:tracks:stats Bumping stat error:unhandled-js-errors` is logged in the console.
10. Reload the page and search `http://pixel.wp.com/g.gif?v=wpcom-no-pv&x_woocommerce-error` in the network tab to confirm the stats are sent to the server. (It would take hours to days to see the stats in the server)
11. Run `git cherry-pick 701c795bf16874bd00b4739e871ef1da945ffa83` (701c795bf16874bd00b4739e871ef1da945ffa83) to simulate the error while rendering the component.
12. Run `pnpm --filter='@woocommerce/plugin-woocommerce' build`
13. Reload the page and search `http://pixel.wp.com/g.gif?v=wpcom-no-pv&x_woocommerce-error` in the network tab to confirm the stats are sent to the server.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
